src/Compilers/CSharp/Portable/Symbols/Metadata/PE/TupleTypeDecoder.cs

```diff
@@ -0,0 +1,233 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Roslyn.Utilities;
+using System;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Reflection.Metadata;
+
+namespace Microsoft.CodeAnalysis.CSharp.Symbols.Metadata.PE
+{
+    internal struct TupleTypeDecoder
+    {
+        private readonly AssemblySymbol _containingAssembly;
+        private readonly ImmutableArray<string> _elementNames;
+        // Keep track of how many names we've "used" during decoding. Starts at
+        // the back of the array and moves forward.
+        private int _namesIndex;
+
+        public TupleTypeDecoder(ImmutableArray<string> elementNames, AssemblySymbol containingAssembly)
+        {
+            _containingAssembly = containingAssembly;
+            _elementNames = elementNames;
+            _namesIndex = elementNames.IsDefault ? 0 : elementNames.Length;
+        }
+
+        public static TypeSymbol DecodeTupleTypesIfApplicable(
+            TypeSymbol metadataType,
+            EntityHandle targetSymbolToken,
+            PEModuleSymbol containingModule)
+        {
+            Debug.Assert((object)metadataType != null);
+
+            ImmutableArray<string> elementNames;
+            // First check if the type is an interface. Interface implementations
+            // currently don't record tuple names in an attribute
+            if (metadataType.IsInterfaceType())
+            {
+                elementNames = default(ImmutableArray<string>);
+                var decoder = new TupleTypeDecoder(elementNames, containingModule.ContainingAssembly);
+                return decoder.DecodeType(metadataType);
+            }
+            if (containingModule.Module.HasTupleElementNamesAttribute(targetSymbolToken, out elementNames))
+            {
+                var decoder = new TupleTypeDecoder(elementNames, containingModule.ContainingAssembly);
+                return decoder.DecodeType(metadataType);
```

Comment:
should we throw here if not all name were consumed? Clearly there is something wrong with the metadata.